src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,317 @@
 {
-    internal static class RemoteHostAssetSerialization
+    /// <summary>
+    /// A sentinel byte we place between messages.  Ensures we can detect when something has gone wrong as soon as
+    /// possible. Note: the value we pick is neither ascii nor extended ascii.  So it's very unlikely to appear
+    /// accidentally.
+    /// </summary>
+    private const byte MessageSentinelByte = 0b10010000;
+
+    private static readonly ObjectPool<SerializableBytes.ReadWriteStream> s_streamPool = new(SerializableBytes.CreateWritableStream);
+
+    private static readonly UnboundedChannelOptions s_channelOptions = new()
     {
-        public static async ValueTask WriteDataAsync(
-            Stream stream,
-            Dictionary<Checksum, object> assetMap,
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
+        // We have a single task reading the data from the channel and writing it to the pipe.  This option
+        // allows the channel to operate in a more efficient manner knowing it won't have to sychronize data
+        // for multiple readers.
+        SingleReader = true,
+
+        // Currently we only have a single writer writing to the channel when we call FindAllAssetsAsync.
+        // However, we could change this in the future to allow the search to happen in parallel.
+        SingleWriter = true,
+    };
+
+    public static async ValueTask WriteDataAsync(
+        PipeWriter pipeWriter,
+        AssetPath assetPath,
+        ReadOnlyMemory<Checksum> checksums,
+        SolutionAssetStorage.Scope scope,
+        ISerializerService serializer,
+        CancellationToken cancellationToken)
+    {
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(s_channelOptions);
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe. Capture-free version is only available on netcore unfortunately.
+#if NET
+        using var _ = cancellationToken.Register(
+            static (obj, cancellationToken) => ((ChecksumChannel)obj!).Writer.TryComplete(new OperationCanceledException(cancellationToken)),
+            state: channel);
+#else
+        using var _ = cancellationToken.Register(
+            () => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
+#endif
+
+        // Spin up a task to go search for all the requested checksums, adding results to the channel.
+        var findAssetsTask = FindAllAssetsAsync(assetPath, checksums, scope, channel, cancellationToken);
```

Comment:
>FindAllAssetsAsync [](http://example.com/codeflow?start=29&length=18)

nit: These names were a bit confusing to me. Something like FindAssetsFromScopeAndWriteToChannelAsync and ReadAssetsFromChannelAndWriteToPipeAsync might have been a bit clearer to me

Reply:
sure.